extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/src/main/java/org/eclipse/smarthome/binding/bluetooth/bluegiga/BlueGigaBluetoothDevice.java

```diff
@@ -0,0 +1,431 @@
+                    break;
+                default:
+                    break;
+            }
+
+            return;
+        }
+
+        if (event instanceof BlueGigaConnectionStatusEvent) {
+            BlueGigaConnectionStatusEvent connectionEvent = (BlueGigaConnectionStatusEvent) event;
+
+            // Check if this is addressed to this device
+            if (!address.equals(new BluetoothAddress(connectionEvent.getAddress()))) {
+                return;
+            }
+
+            // If we're connected, then remember the connection handle
+            if (connectionEvent.getFlags().contains(ConnectionStatusFlag.CONNECTION_CONNECTED)) {
+                connectionState = ConnectionState.CONNECTED;
+                connection = connectionEvent.getConnection();
+            }
+
+            if (connectionEvent.getFlags().contains(ConnectionStatusFlag.CONNECTION_CONNECTED)) {
+                notifyListeners(BluetoothEventType.CONNECTION_STATE,
+                        new BluetoothConnectionStatusNotification(connectionState));
+            }
+
+            return;
+        }
+
+        if (event instanceof BlueGigaDisconnectedEvent) {
+            BlueGigaDisconnectedEvent disconnectedEvent = (BlueGigaDisconnectedEvent) event;
+
+            // If this is not our connection handle then ignore.
+            if (connection != disconnectedEvent.getConnection()) {
+                return;
+            }
+
+            connectionState = ConnectionState.DISCONNECTED;
+            connection = -1;
+            notifyListeners(BluetoothEventType.CONNECTION_STATE,
+                    new BluetoothConnectionStatusNotification(connectionState));
+
+            return;
+        }
+
+        if (event instanceof BlueGigaAttributeValueEvent) {
+            // A read request has completed - update the characteristic
+            BlueGigaAttributeValueEvent valueEvent = (BlueGigaAttributeValueEvent) event;
+
+            BluetoothCharacteristic characteristic = getCharacteristicByHandle(valueEvent.getAttHandle());
+            if (characteristic == null) {
+                logger.debug("BlueGiga didn't find characteristic for event {}", event);
+            }
+
+            // If this is the characteristic we were reading, then send a read completion
+            if (procedureProgress == BlueGigaProcedure.CHARACTERISTIC_READ && procedureCharacteristic != null
+                    && procedureCharacteristic.getHandle() == valueEvent.getAttHandle()) {
+                procedureProgress = BlueGigaProcedure.NONE;
+                procedureCharacteristic = null;
+                notifyListeners(BluetoothEventType.CHARACTERISTIC_READ_COMPLETE, procedureCharacteristic,
+                        BluetoothCompletionStatus.SUCCESS);
+            }
+
+            // Notify the user of the updated value
+            if (characteristic != null) {
+                notifyListeners(BluetoothEventType.CHARACTERISTIC_UPDATED, procedureCharacteristic);
```

Comment:
@cdjackson this call pass null to the listener very often, can you agree with me that maybe passing `characteristic` to the listeners might make more sense, or do I overlook some detail here?

Reply:
Yes - you are right - this should be ```characteristic```.

Reply:
Hm, but why does the if check for `procedureCharacteristic`? Should that be changed to `characteristic` as well? Or should it be an additional check?